core/src/main/java/org/ethereum/beacon/core/state/BeaconStateImpl.java

```diff
@@ -0,0 +1,428 @@
+        new ArrayList<>(state.getLatestVdfOutputs()),
+        state.getShardCommitteesAtSlots().stream().map(ArrayList::new).collect(Collectors.toList()),
+
+        new ArrayList<>(state.getCustodyChallenges()),
+
+        state.getPreviousJustifiedSlot(),
+        state.getJustifiedSlot(),
+        state.getJustificationBitfield(),
+        state.getFinalizedSlot(),
+
+        new ArrayList<>(state.getLatestCrosslinks()),
+        new ArrayList<>(state.getLatestBlockRoots()),
+        new ArrayList<>(state.getLatestPenalizedExitBalances()),
+        new ArrayList<>(state.getLatestAttestations()),
+        new ArrayList<>(state.getBatchedBlockRoots()),
+
+        state.getLatestDepositRoot(),
+        new ArrayList<>(state.getDepositRootVotes())
+    );
+  }
+      
+  private BeaconStateImpl(
+      UInt64 slot,
+      UInt64 genesisTime,
+      ForkData forkData,
+      List<ValidatorRecord> validatorRegistry,
+      List<UInt64> validatorBalances,
+      UInt64 validatorRegistryLatestChangeSlot,
+      UInt64 validatorRegistryExitCount,
+      Hash32 validatorRegistryDeltaChainTip,
+      List<Hash32> latestRandaoMixes,
+      List<Hash32> latestVdfOutputs,
+      List<List<ShardCommittee>> shardCommitteesAtSlots,
+      List<CustodyChallenge> custodyChallenges,
+      UInt64 previousJustifiedSlot,
+      UInt64 justifiedSlot,
+      UInt64 justificationBitfield,
+      UInt64 finalizedSlot,
+      List<CrosslinkRecord> latestCrosslinks,
+      List<Hash32> latestBlockRoots,
+      List<UInt64> latestPenalizedExitBalances,
+      List<PendingAttestationRecord> latestAttestations,
+      List<Hash32> batchedBlockRoots,
+      Hash32 latestDepositRoot,
+      List<DepositRootVote> depositRootVotes) {
+    this.slot = slot;
+    this.genesisTime = genesisTime;
+    this.forkData = forkData;
+    this.validatorRegistry = validatorRegistry;
+    this.validatorBalances = validatorBalances;
+    this.validatorRegistryLatestChangeSlot = validatorRegistryLatestChangeSlot;
+    this.validatorRegistryExitCount = validatorRegistryExitCount;
+    this.validatorRegistryDeltaChainTip = validatorRegistryDeltaChainTip;
+    this.latestRandaoMixes = latestRandaoMixes;
+    this.latestVdfOutputs = latestVdfOutputs;
+    this.shardCommitteesAtSlots = shardCommitteesAtSlots;
+    this.custodyChallenges = custodyChallenges;
+    this.previousJustifiedSlot = previousJustifiedSlot;
+    this.justifiedSlot = justifiedSlot;
+    this.justificationBitfield = justificationBitfield;
+    this.finalizedSlot = finalizedSlot;
+    this.latestCrosslinks = latestCrosslinks;
+    this.latestBlockRoots = latestBlockRoots;
+    this.latestPenalizedExitBalances = latestPenalizedExitBalances;
+    this.latestAttestations = latestAttestations;
+    this.batchedBlockRoots = batchedBlockRoots;
+    this.latestDepositRoot = latestDepositRoot;
+    this.depositRootVotes = depositRootVotes;
+
+    validate();
+  }
+
+  @Override
+  public BeaconState validate() {
```

Comment:
IMO, for `MutableBeaconState` it's better to run validation in each setter separately. Client might just forget to call `validate()` after all the changes.

Reply:
Yep, I don't like this either too much. 
I'm not sure the State can be consistent on each `setXXX` call

Reply:
For sure, it can't. But you may check only an argument that is changed by particular setter. I.e. check `XXX` in case of `setXXX`.